### ISSUE TITLE
Adding auth header causes React error

### ISSUE THREAD

**Issue by rallieon** [coinbase/rest-hooks#1101]
**React version** (e.g., 16.8.5)
18.0.0-0

**Concurrent mode** yes/no
Not sure

**Rest Hooks version** (e.g., 4.5.9)
5.3.3
@rest-hooks/rest version is 2.1.4

**Describe the bug**
I am trying to create a base resource to perform authentication using a bearer token like the example [here](https://resthooks.io/docs/guides/auth#auth-headers-from-react-context). However, I am receiving the following error message:

```
Warning: React has detected a change in the order of Hooks called by NewTodo. This will lead to bugs and errors if not fixed. For more information, read the Rules of Hooks: https://reactjs.org/link/rules-of-hooks
```

I'm wondering if I am simply performing the configuration wrong. Is there an example application that utilizes the hook pattern inside the `useFetchInit` override?

**To Reproduce**
1. Download the master branch of `rest-hooks`. 
2. Modify the package.json of the Todo example application to the following:
```
{
  "name": "todo-app",
  "version": "0.1.0",
  "description": "todo-app - A rest hooks example",
  "scripts": {
    "lint": "eslint src --ext .ts,.tsx",
    "format": "npm run lint --fix",
    "test:type": "tsc",
    "start": "webpack serve --mode=development --port=3001 --host 0.0.0.0",
    "start:prod": "serve dist",
    "build:browser": "webpack --mode=production",
    "build:analyze": "webpack --mode=production --env analyze",
    "build:profile": "webpack --mode=production --env profile",
    "test:pkg": "webpack --env check=nobuild"
  },
  "keywords": [
    "anansi"
  ],
  "main": "src/index.ts",
  "types": "src/index.ts",
  "private": true,
  "license": "Apache-2.0",
  "devDependencies": {
    "@anansi/babel-preset": "2.7.8",
    "@anansi/browserslist-config": "1.3.3",
    "@anansi/eslint-plugin": "0.11.21",
    "@anansi/webpack-config": "8.0.3",
    "@babel/core": "7.15.0",
    "@linaria/babel-preset": "3.0.0-beta.4",
    "@linaria/core": "3.0.0-beta.4",
    "@linaria/react": "3.0.0-beta.7",
    "@linaria/shaker": "3.0.0-beta.7",
    "@rest-hooks/rest": "2.1.4",
    "@types/react-dom": "17.0.9",
    "react-refresh": "^0.11.0-0",
    "rest-hooks": "5.3.3",
    "serve": "12.0.0",
    "webpack": "5.48.0",
    "webpack-cli": "4.7.2",
    "webpack-dev-server": "3.11.2"
  },
  "dependencies": {
    "webpack-cli": "^4.7.2"
  }
}

```
3. Change the ToDoResource.ts to the following:
```
import { Resource } from '@rest-hooks/rest';
import { createContext, useContext } from 'react';

import PlaceholderBaseResource from './PlaceholderBaseResource';

export const AuthContext = createContext('validjwt');

export default class TodoResource extends PlaceholderBaseResource {
  readonly userId: number = 0;
  readonly title: string = '';
  readonly completed: boolean = false;

  static urlRoot = 'https://jsonplaceholder.typicode.com/todos';

  static useFetchInit = (init: RequestInit) => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const accessToken = useContext(AuthContext);
    // eslint-disable-next-line no-console
    console.log(accessToken);
    return {
      ...init,
      headers: {
        ...init.headers,
        Authorization: `Bearer ${accessToken}`,
      },
    };
  };

  static partialUpdate<T extends typeof Resource>(this: T) {
    return super.partialUpdate().extend({
      schema: this,
      optimisticUpdate: optimisticPartial,
    });
  }

  static create<T extends typeof Resource>(this: T) {
    return super.create().extend({
      schema: this,
      optimisticUpdate: optimisticCreate,
      update: (newResourceId: string) => ({
        [this.list().key({})]: (resourceIds: string[] = []) => [
          ...resourceIds,
          newResourceId,
        ],
      }),
    });
  }
}

const optimisticPartial = (params: any, body: any) => ({
  id: params.id,
  ...body,
});

const optimisticCreate = (_: any, body: any) => body;

```
4. Load application via start command and try to click the checkbox of one of the todo items.
5. Look in the console for the error.

**Expected behavior**
The expectation is not to receive the warning from React.

**Additional context**
When creating a new todo item by typing in the textbox and hitting 'Enter' I receive the following error:

```
Warning: React has detected a change in the order of Hooks called by CacheProvider. This will lead to bugs and errors if not fixed. For more information, read the Rules of Hooks: https://reactjs.org/link/rules-of-hooks
```

and

```
Warning: Do not call Hooks inside useEffect(...), useMemo(...), or other built-in Hooks. You can only call Hooks at the top level of your React function. For more information, see https://reactjs.org/link/rules-of-hooks
CacheProvider@http://localhost:3001/App.js:3791:89
RootProvider@http://localhost:3001/App.js:45194:22
```

I have attached screenshots of the diff of the files I changed:

package.json:

![image](https://user-images.githubusercontent.com/638224/128649393-da92605d-9d12-4269-b446-72bb3470d7fb.png)

TodoResource.ts
![image](https://user-images.githubusercontent.com/638224/128649412-4793224f-bb1b-4a98-a886-5389123a7f26.png)

**Comment by ntucker** [comment]
Thanks for the detailed report - it will make diagnosing this much easier. Will look into this later today.

**Comment by ntucker** [comment]
Fixed in [@rest-hooks/rest@2.2.0](https://www.npmjs.com/package/@rest-hooks/rest/v/2.2.0) @rallieon

**Comment by rallieon** [comment]
@ntucker Thank you!
